File: 61.rotate-list.java

/*
 * @lc app=leetcode id=61 lang=java
 *
 * [61] Rotate List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {

        if (k<=0|| head == null || head.next == null) {
            return head;
        }
        int i = 0;

        int len = 1;
        ListNode temp = head;
        while (temp.next != null) {
            len++;
            temp = temp.next;
        }
        // System.out.println(len);
        
        k = k % len;
        if (k <= 0) {
            return head;
        }
        // System.out.println(len - k);

        i = 1;
        ListNode prevLast = head;
        while (i < len - k) {

            prevLast = prevLast.next;
            i++;

        }
        ListNode newHead = prevLast.next;
        prevLast.next = null;
        temp.next = head;

//         while(head!=null){
// System.out.println(head.val);
// head=head.next;
//         }

        return newHead;

    }

}
// @lc code=end

File: 61.rotate-list.cpp
/*
 * @lc app=leetcode id=61 lang=cpp
 *
 * [61] Rotate List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {

        ListNode* ptr = head;
        if(head==NULL){
            return head;
        }
        if(k==0){
            return head;
        }

        int len=1;
        while(ptr->next!=NULL){
            len++;
            ptr=ptr->next;
        }

        int r = k%len;
        ptr->next = head;
        ListNode* ptr2 = head;
        int i=1;
        while(i<len-r){
            ptr2=ptr2->next;
            i++;
        }
        ptr = ptr2->next;
        ptr2->next = NULL;
        return ptr;

    }
};
// @lc code=end

